mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp

```diff
@@ -1818,6 +1818,11 @@ struct RemoveOutsDependency : public OpRewritePattern<GenericOp> {
         if (sparse_tensor::getSparseTensorEncoding(operandVal.getType()))
           continue;
 
+        // If outs is wired from a block argument, keep the dependency to
+        // prevent the argument from being optimized away.
```

Comment:
This comment as-is is weird: why would we not want a block argument to be optimized away?

Reply:
The pattern rewriter actually breaks the original dependency for an argument used for `outs`. (Assuming that a tensor *can* used for output after some lowering, we need to keep this dependency.)

Reply:
Sorry, you'll need to elaborate: this outs isn't actually used, the "we need to keep this dependency as is" isn't clear to me.

Reply:
Let me explain the context, which may give you more info and what is happening here.

At 10,000 feet, I'd like to use a tensor operand as output. Here is an example,

```mlir
func.func @my_add(%a: tensor<2xf32>, %b: tensor<2xf32>, %out: tensor<2xf32> { iree.abi.output : 0} -> tensor<2xf32> {
  %r = linalg.generic ins(%a : tensor<2xf32>, %b : tensor<2xf32>) outs(%out) {
    // add op
    ...
  } -> tensor<2xf32>
  return %r : tensor<2xf32>
}
```

Here with `iree.abi.output`, we say that `%out` is used for output. So the function does not allocate buffer for `%r` and uses the caller-allocated storage for `%out` for the output storage.

But the rewriter breaks the existing dependency and blindly introduces `tensor.empty()` which makes `%out` used, so it gets removed later.

Hope this helps.

Reply:
> At 10,000 feet, I'd like to use a tensor operand as output.

To begin with, this is ill-defined: you mean as an "init" of the linalg.generic right?

> Here with iree.abi.output, we say that %out is used for output. So the function does not allocate buffer for %r and uses the caller-allocated storage for %out for the output storage.

Seems like a nice contract to provide to a bufferization optimization, but fundamentally that seems like it can't rely on the `init` of the linalg generic being "by chance" the one you want: the bufferization should recover.

> But the rewriter breaks the existing dependency and blindly introduces tensor.empty() which makes %out unused, so it gets removed later.

Is your function "private"? Are you running some global optimization pass? If it is an upstream pass, yeah it can't know about your ABI attribute and you need another mechanism (mark these with ops in the function instead? Like `iree.abi_output(%arg0)` ?

Reply:
> To begin with, this is ill-defined: you mean as an "init" of the linalg.generic right?
>Or are you referring to the IREE ABI attribute? (which really can't have much bearing upstream...)

I may say that the logic around the tensor empty and linalg init is difficult to grasp for me too. I described the context with the IREE attribute, but it would be great we can have a solution to use a function operand as output.

My change here is pretty vanilla as a preparation step to take next steps; the main question here for me was "Can I use a block argument for `outs`? And, actually the change was suggested by @MaheshRavishankar in a different thread. I also thought it is okay.

Do you think it is ill-defined because a tensor is not supposed to be written? I had the same question for the tensor from `tensor.empty()`.

One solution IREE used is converting the ABI operands into `iree.hal.buffer` and supports importing and exporting. Generalizing this might be a good solution to have a tensor reference semantic.

Another solution is to have memref from the very early stage of the compilation process. But there might be people who prefer to optimizations with the tensor representation before going to the stages with memref.

> Seems like a nice contract to provide to a bufferization optimization, but fundamentally that seems like it can't rely on the init of the linalg generic being "by chance" the one you want: the bufferization should recover.

What do you mean "bufferization should recover"?

Regarding `tensor.empty` used as "init" is also not very clear about the storage. Actually, tensor as value is confusing without having a reference or pointer to a storage. I thought tensor.empty follows an allocation semantic, but we allow CSE for it if the shape is the same.

> Is your function "private"? Are you running some global optimization pass? If it is an upstream pass, yeah it can't know about your ABI attribute and you need another mechanism (mark these with ops in the function instead? Like iree.abi_output(%arg0) ?

Yes, the function is private and there is a pass (I created, but not yet reviewed) that finds the function return value to the tensor empty and replace the tensor empty value with the argument with the attribute.

For now, the problem is at the ABI level, but we can generalize the situation as the following questions:

- "How can we use a tensor operand as output?" 
  - Is it ill-formed?
  - If so, do we want to introduce an attribute to mark a tensor writable?
- Do we want to introduce type for tensor reference?
- Do we want to use memref directly along with tensor? (Sounds feasible.)

Thanks for your time to raise the questions! It is good to have a discussion here.

@MaheshRavishankar feel free to chime in and give a comment!

Reply:
>  "Can I use a block argument for outs?

You can (the IR is valid and has a well defined semantics), but you can't guarantee it'll stay there (this transformation that you want to block is a valid one).
Semantically there is nothing "special" about a block argument, so it's hard for me to understand what would justify this change right now.

> Do you think it is ill-defined because a tensor is not supposed to be written?

Yes, tensors are immutable object: the `outs` in linalg-on-tensor refers to two things as far as I remember: 
1) The shape of the output, in case of dynamic output.
2) The initial value in case of reduction iterators.

Here it is unused and can be eliminated: this seems like actually a desirable feature and optimization that you would be removing here!

> I had the same question for the tensor from tensor.empty().

What is your question on tensor.empty()? I see it as a tensor with a "poison value", only the shape can be used.

> Another solution is to have memref from the very early stage of the compilation process. But there might be people who prefer to optimizations with the tensor representation before going to the stages with memref.

There are many ways to implement a well defined and robust solution actually, let me make-up one for example:

IREE could express its ABI with memrefs while still using tensors in the body by converting to tensor there:

```
func.func @my_add(%a_: memref<2xf32>, %b_: memref<2xf32>, %out_: memref<2xf32>) {
  %a = iree.to_tensor %a_
  %b = iree.to_tensor %b_
  %out = iree.to_tensor %out_
  
  %r = linalg.generic ins(%a : tensor<2xf32>, %b : tensor<2xf32>) outs(%out) {
    // add op
    ...
  } -> tensor<2xf32>
  
  iree.to_memref %r, %out_ : tensor<2xf32>, memref<2xf32>
  return
}
```

Of course we wouldn't preserve the `outs` of the linalg generic (because... why?) but the `iree.to_memref` keeps it alive (the `%out` won't get removed).

> What do you mean "bufferization should recover"?

I mean that in the example above, after removing `%out` it is a matter of optimization in the buffer allocation algorithm to be able to reuse the buffer from `%out_` instead of inserting a copy.


> "How can we use a tensor operand as output?"
> Is it ill-formed?

I'm saying that the question itself is ill-formed: tensors are immutable value-based object.

> If so, do we want to introduce an attribute to mark a tensor writable?

We instead went with another type: `memref` :)

> Do we want to introduce type for tensor reference?

I can't comment on this: what would be the proposed semantic? which op would support this? (and how) 
It all depends what you mean by "tensor reference" actually.

Reply:
> Yes, tensors are immutable object: the outs in linalg-on-tensor refers to two things as far as I remember:
>
> The shape of the output, in case of dynamic output.
> The initial value in case of reduction iterators.

Thanks a lot @joker-eph for this. I think I now understand the tensor emtpy semantic much better.

Based on the semantic, linking the information from the output storage through the DPS doesn't seem to be a robust way. I really like the explicit way using an operation. Since tensor is supposed to be immutable, let me try to preserve the property. I was hesitant to introduce memref before bufferization, but memref + explicit operation seems to be the cleanest way without introducing some hidden dependency on how things should work implicitly.